File: src/services/github.ts
import { Effect } from "effect";
import type { FlakeInput, GitHubCommit, UpdateStatus } from "../types";
import { GitHubApiError, GitHubInputError, GitHubRateLimitError } from "../types/errors";

const API_DELAY_MS = Number(process.env.MELT_API_DELAY_MS) || 0;
const delay = (ms: number): Effect.Effect<void> =>
	Effect.promise(() => new Promise((resolve) => setTimeout(resolve, ms)));

// Get GitHub token from environment for higher rate limits
// Supports GITHUB_TOKEN, GH_TOKEN, and GITHUB_PAT
const GITHUB_TOKEN = process.env.GITHUB_TOKEN || process.env.GH_TOKEN || process.env.GITHUB_PAT;

interface GitHubAPICommit {
	sha: string;
	commit: {
		message: string;
		author: {
			name: string;
			date: string;
		};
	};
	html_url: string;
}

/**
 * Get headers for GitHub API requests
 */
function getGitHubHeaders(): Record<string, string> {
	const headers: Record<string, string> = {
		Accept: "application/vnd.github.v3+json",
		"User-Agent": "melt-tui",
	};

	if (GITHUB_TOKEN) {
		headers.Authorization = `token ${GITHUB_TOKEN}`;
	}

	return headers;
}

/**
 * Format a date string as short relative time (e.g., "2d ago")
 */
function formatShortRelativeTime(dateStr: string): string {
	const date = new Date(dateStr);
	const now = new Date();
	const diffMs = now.getTime() - date.getTime();
	const diffSecs = Math.floor(diffMs / 1000);

	if (diffSecs < 60) {
		return "just now";
	}

	const TIME_UNITS = [
		{ seconds: 365 * 24 * 60 * 60, short: "y" },
		{ seconds: 30 * 24 * 60 * 60, short: "mo" },
		{ seconds: 7 * 24 * 60 * 60, short: "w" },
		{ seconds: 24 * 60 * 60, short: "d" },
		{ seconds: 60 * 60, short: "h" },
		{ seconds: 60, short: "m" },
	];

	for (const unit of TIME_UNITS) {
		if (diffSecs >= unit.seconds) {
			const count = Math.floor(diffSecs / unit.seconds);
			// For months and beyond, use date format
			if (unit.short === "mo" || unit.short === "y") {
				return date.toLocaleDateString("en-US", {
					month: "short",
					day: "numeric",
				});
			}
			return `${count}${unit.short} ago`;
		}
	}

	return "just now";
}

/**
 * Make a GitHub API request with error handling
 */
const fetchGitHub = (url: string): Effect.Effect<Response, GitHubRateLimitError | GitHubApiError> =>
	Effect.gen(function* () {
		if (API_DELAY_MS > 0) yield* delay(API_DELAY_MS);

		const response = yield* Effect.tryPromise({
			try: () =>
				fetch(url, {
					headers: getGitHubHeaders(),
				}),
			catch: () =>
				new GitHubApiError({
					status: 0,
					statusText: "Network error",
					url,
				}),
		});

		if (!response.ok) {
			if (response.status === 403) {
				const remaining = response.headers.get("X-RateLimit-Remaining");
				if (remaining === "0") {
					const resetHeader = response.headers.get("X-RateLimit-Reset");
					const resetAt = resetHeader ? new Date(Number(resetHeader) * 1000) : undefined;
					return yield* Effect.fail(new GitHubRateLimitError({ resetAt }));
				}
			}
			return yield* Effect.fail(
				new GitHubApiError({
					status: response.status,
					statusText: response.statusText,
					url,
				}),
			);
		}

		return response;
	});

/**
 * Fetch commits starting from a specific revision (inclusive) and going back in history
 */
const getCommitsFromRev = (
	owner: string,
	repo: string,
	fromRev: string,
	limit: number = 50,
): Effect.Effect<GitHubCommit[], GitHubRateLimitError | GitHubApiError> =>
	Effect.gen(function* () {
		const commits: GitHubCommit[] = [];
		let page = 1;
		const perPage = Math.min(limit, 100);

		while (commits.length < limit && page <= 3) {
			const url = `https://api.github.com/repos/${owner}/${repo}/commits?sha=${fromRev}&per_page=${perPage}&page=${page}`;

			const response = yield* fetchGitHub(url);
			const data = (yield* Effect.tryPromise({
				try: () => response.json() as Promise<GitHubAPICommit[]>,
				catch: () =>
					new GitHubApiError({
						status: 0,
						statusText: "Failed to parse JSON",
						url,
					}),
			})) as GitHubAPICommit[];

			if (data.length === 0) break;

			for (const commit of data) {
				if (commits.length >= limit) break;

				const message = commit.commit.message.split("\n")[0];

				commits.push({
					sha: commit.sha,
					shortSha: commit.sha.substring(0, 7),
					message: message ?? "",
					author: commit.commit.author.name,
					date: formatShortRelativeTime(commit.commit.author.date),
					url: commit.html_url,
				});
			}

			page++;
		}

		return commits;
	});

/**
 * Check if an update is available for a flake input (GitHub only)
 */
const checkForUpdate = (input: FlakeInput): Effect.Effect<UpdateStatus, never> =>
	Effect.gen(function* () {
		if (!input.owner || !input.repo) {
			return {
				commitsBehind: 0,
				loading: false,
				error: "Missing owner or repo",
			};
		}

		const result = yield* githubService
			.getCommitsSinceRev(input.owner, input.repo, input.rev, input.ref)
			.pipe(Effect.either);

		if (result._tag === "Left") {
			return {
				commitsBehind: 0,
				loading: false,
				error: result.left.message,
			};
		}

		return {
			commitsBehind: result.right.length,
			loading: false,
		};
	});

export const githubService = {
	/**
	 * Fetch commits from GitHub since a specific revision
	 * Returns commits from HEAD back to (but not including) the specified rev
	 */
	getCommitsSinceRev: (
		owner: string,
		repo: string,
		sinceRev: string,
		ref?: string,
	): Effect.Effect<GitHubCommit[], GitHubRateLimitError | GitHubApiError> =>
		Effect.gen(function* () {
			const commits: GitHubCommit[] = [];
			let page = 1;
			const perPage = 100;
			let foundRev = false;

			while (!foundRev && page <= 5) {
				// Use sha param to specify branch/ref if provided
				const shaParam = ref ? `&sha=${encodeURIComponent(ref)}` : "";
				const url = `https://api.github.com/repos/${owner}/${repo}/commits?per_page=${perPage}&page=${page}${shaParam}`;

				const response = yield* fetchGitHub(url);
				const data = (yield* Effect.tryPromise({
					try: () => response.json() as Promise<GitHubAPICommit[]>,
					catch: () =>
						new GitHubApiError({
							status: 0,
							statusText: "Failed to parse JSON",
							url,
						}),
				})) as GitHubAPICommit[];

				if (data.length === 0) break;

				for (const commit of data) {
					// Stop when we find the locked revision
					if (commit.sha === sinceRev || commit.sha.startsWith(sinceRev)) {
						foundRev = true;
						break;
					}

					const message = commit.commit.message.split("\n")[0];

					commits.push({
						sha: commit.sha,
						shortSha: commit.sha.substring(0, 7),
						message: message ?? "",
						author: commit.commit.author.name,
						date: formatShortRelativeTime(commit.commit.author.date),
						url: commit.html_url,
					});
				}

				page++;
			}

			return commits;
		}),

	/**
	 * Fetch commit history around a locked revision
	 * Returns commits ahead (newer), the locked commit, and commits behind (older)
	 */
	getCommitHistory: (
		owner: string,
		repo: string,
		lockedRev: string,
		ref?: string,
		behindLimit: number = 50,
	): Effect.Effect<{ commits: GitHubCommit[]; lockedIndex: number }, GitHubRateLimitError | GitHubApiError> =>
		Effect.gen(function* () {
			// Fetch commits ahead of the locked rev (newer commits)
			const commitsAhead = yield* githubService.getCommitsSinceRev(owner, repo, lockedRev, ref);

			// Fetch commits from the locked rev and older
			const commitsFromLocked = yield* getCommitsFromRev(owner, repo, lockedRev, behindLimit + 1);

			// Mark the first commit (the locked one) with isLocked flag
			const lockedCommit = commitsFromLocked[0];
			if (lockedCommit) {
				lockedCommit.isLocked = true;
			}

			// Combine: ahead commits + locked commit + behind commits
			const allCommits = [...commitsAhead, ...commitsFromLocked];
			const lockedIndex = commitsAhead.length;

			return { commits: allCommits, lockedIndex };
		}),

	/**
	 * Get changelog for a flake input (only works for GitHub inputs)
	 */
	getChangelog: (
		input: FlakeInput,
	): Effect.Effect<
		{ commits: GitHubCommit[]; lockedIndex: number },
		GitHubRateLimitError | GitHubApiError | GitHubInputError
	> =>
		Effect.gen(function* () {
			if (input.type !== "github" || !input.owner || !input.repo) {
				return yield* Effect.fail(
					new GitHubInputError({
						inputName: input.name,
						reason: "Changelog is only available for GitHub inputs",
					}),
				);
			}

			return yield* githubService.getCommitHistory(input.owner, input.repo, input.rev, input.ref);
		}),

	/**
	 * Check for updates on multiple inputs in parallel
	 */
	checkForUpdates: (
		inputs: FlakeInput[],
		onStatusChange?: (name: string, status: UpdateStatus) => void,
	): Effect.Effect<Map<string, UpdateStatus>> =>
		Effect.gen(function* () {
			const results = new Map<string, UpdateStatus>();

			// Filter to only GitHub inputs
			const githubInputs = inputs.filter((input) => input.type === "github" && input.owner && input.repo);

			for (const input of githubInputs) {
				const loadingStatus: UpdateStatus = {
					commitsBehind: 0,
					loading: true,
				};
				results.set(input.name, loadingStatus);
				onStatusChange?.(input.name, loadingStatus);
			}

			// Batch to avoid rate limits (Unauth: 60/hr, Auth: 5000/hr)
			const BATCH_SIZE = githubService.hasGitHubToken() ? 10 : 2;

			for (let i = 0; i < githubInputs.length; i += BATCH_SIZE) {
				const batch = githubInputs.slice(i, i + BATCH_SIZE);
				yield* Effect.forEach(
					batch,
					(input) =>
						Effect.gen(function* () {
							const status = yield* checkForUpdate(input);
							results.set(input.name, status);
							onStatusChange?.(input.name, status);
						}),
					{ concurrency: "unbounded" },
				);
			}

			return results;
		}),

	hasGitHubToken: (): boolean => !!GITHUB_TOKEN,
};

export type GitHubService = typeof githubService;

File: src/service/git/github.rs
use chrono::Utc;
use reqwest::Client;
use serde::Deserialize;
use tracing::warn;

use crate::error::GitError;
use crate::model::{ChangelogData, Commit, GitInput};

pub struct GithubClient<'a> {
    pub client: &'a Client,
    pub token: Option<&'a str>,
}

impl<'a> GithubClient<'a> {
    pub async fn check_updates(&self, input: &GitInput) -> Result<usize, GitError> {
        let branch = input.reference.as_deref().unwrap_or("HEAD");
        let url = format!(
            "https://api.github.com/repos/{}/{}/compare/{}...{}",
            input.owner, input.repo, input.rev, branch
        );

        let mut req = self.client.get(&url);
        if let Some(token) = self.token {
            req = req.header("Authorization", format!("Bearer {}", token));
        }

        let resp = req
            .send()
            .await
            .map_err(|e| GitError::NetworkError(e.to_string()))?;

        let status = resp.status();

        if status.as_u16() == 403 || status.as_u16() == 429 {
            let remaining = resp
                .headers()
                .get("x-ratelimit-remaining")
                .and_then(|v| v.to_str().ok())
                .and_then(|v| v.parse::<u32>().ok())
                .unwrap_or(0);

            if remaining == 0 {
                warn!(input = %input.name, "GitHub API rate limit exceeded");
                return Err(GitError::NetworkError(
                    "GitHub API rate limit exceeded. Set GITHUB_TOKEN for higher limits."
                        .to_string(),
                ));
            }
        }

        if !status.is_success() {
            return Err(GitError::ApiFallbackRequired);
        }

        #[derive(Deserialize)]
        struct CompareResponse {
            ahead_by: usize,
        }

        let data: CompareResponse = resp
            .json()
            .await
            .map_err(|e| GitError::NetworkError(e.to_string()))?;

        Ok(data.ahead_by)
    }

    pub async fn get_changelog(&self, input: &GitInput) -> Result<ChangelogData, GitError> {
        let branch = input.reference.as_deref().unwrap_or("HEAD");

        let url = format!(
            "https://api.github.com/repos/{}/{}/commits?sha={}&per_page=100",
            input.owner, input.repo, branch
        );

        let mut req = self.client.get(&url);
        if let Some(token) = self.token {
            req = req.header("Authorization", format!("Bearer {}", token));
        }

        let resp = req
            .send()
            .await
            .map_err(|e| GitError::NetworkError(e.to_string()))?;

        let status = resp.status();

        if status.as_u16() == 403 || status.as_u16() == 429 {
            let remaining = resp
                .headers()
                .get("x-ratelimit-remaining")
                .and_then(|v| v.to_str().ok())
                .and_then(|v| v.parse::<u32>().ok())
                .unwrap_or(0);

            if remaining == 0 {
                return Err(GitError::NetworkError(
                    "GitHub API rate limit exceeded. Set GITHUB_TOKEN for higher limits."
                        .to_string(),
                ));
            }
        }

        if !status.is_success() {
            return Err(GitError::ApiFallbackRequired);
        }

        #[derive(Deserialize)]
        struct GitHubAuthor {
            name: Option<String>,
            date: Option<String>,
        }

        #[derive(Deserialize)]
        struct GitHubCommitData {
            message: String,
            author: Option<GitHubAuthor>,
        }

        #[derive(Deserialize)]
        struct GitHubCommit {
            sha: String,
            commit: GitHubCommitData,
        }

        let commits: Vec<GitHubCommit> = resp
            .json()
            .await
            .map_err(|e| GitError::NetworkError(e.to_string()))?;

        let mut result_commits = Vec::new();
        let mut locked_idx = None;

        for (idx, c) in commits.iter().enumerate() {
            let is_locked = c.sha.starts_with(&input.rev) || c.sha == input.rev;
            if is_locked {
                locked_idx = Some(idx);
            }

            let date = c
                .commit
                .author
                .as_ref()
                .and_then(|a| a.date.as_ref())
                .and_then(|d| chrono::DateTime::parse_from_rfc3339(d).ok())
                .map(|d| d.with_timezone(&Utc))
                .unwrap_or_else(Utc::now);

            let author = c
                .commit
                .author
                .as_ref()
                .and_then(|a| a.name.clone())
                .unwrap_or_else(|| "Unknown".to_string());

            let message = c.commit.message.lines().next().unwrap_or("").to_string();

            result_commits.push(Commit {
                sha: c.sha.clone(),
                message,
                author,
                date,
                is_locked,
            });
        }

        Ok(ChangelogData {
            commits: result_commits,
            locked_idx,
        })
    }
}
